docs/design/sum_types.md

```diff
@@ -0,0 +1,200 @@
+declaration, but gives the author full control over the representation and class
+members.
+
+The ability to create instances of the sum type can be straightforwardly
+emulated with factory functions and static constants, and the internal storage
+layout will presumably involve untagged unions or some other low-level storage
+primitive which hasn't been designed yet, but the key to defining a sum type's
+interface is enabling it to support pattern matching. To do that, the sum type
+has to specify two things:
+
+-   The set of all possible alternatives, including their names and parameter
+    types, so that the compiler can typecheck the `match` body, identify any
+    unreachable `case`s, and determine whether any `case`s are missing.
+-   The algorithm that, given a value of the sum type, determines which
+    alternative is present, and specifies the values of its parameters.
+
+Here's how that would look if `Optional` were defined as a class:
+
+```carbon
+class Optional(T:! Type) {
+  // Factory functions
+  fn Some(value: T) -> Self;
+  let None:! Self;
+
+  private var has_value: bool;
+  private var value: T;
+
+  interface MatchContinuation {
+    let template ReturnType:! Type;
+    fn Some[addr me: Self*](value: T) -> ReturnType;
+    fn None[addr me: Self*]() -> ReturnType;
+  }
+
+  external impl as Match(MatchContinuation) {
+    fn Op[me: Self, Continuation:! MatchContinuation](
+        continuation: Continuation*) -> Continuation.ReturnType {
+      if (me.has_value) {
+        return continuation->Some(me.value);
+      } else {
+        return continuation->None();
+      }
+    }
+  }
+
+  // Operations like destruction, copying, assignment, and comparison are
+  // omitted for brevity.
+}
+```
+
+In this code, `Optional` makes itself available for use in pattern matching by
+declaring that it implements the `Match` interface. `Match` takes an interface
+argument, called the _continuation interface_, which specifies the set of
+possible alternatives by declaring a method for each one. In this case, we pass
+`Optional.MatchContinuation` as the continuation interface.
+
+When compiling a `match` statement, the compiler checks that the type being
+matched implements `Match(C)` for some continuation interface `C`. Then, it
+notionally transforms the `match` body into a class that implements `C`, with
+one method for each `case`, and then passes that to `Match.Op` on the object
+being matched. For example, the `match` statement shown earlier might be
+transformed into:
+
+```carbon
+class __MatchStatementImpl {
+  fn Make() -> Self { return {}; }
+
+  impl as Match(Optional.MatchContinuation) where .ReturnType = () {
+    fn Some(the_value: i32) {
+      Print(the_value);
+    }
+    fn None() {
+      Print("None");
+    }
+  }
+}
+
+my_opt.(Match.Op)(__MatchStatementImpl.Make());
+```
+
+Thus, a `match` statement works by invoking the sum type's `Match.Op` method,
```

Comment:
Is there any guarantee as to how many times `Match.Op` might be called, or in what order? In #2188 I said effectively that we pattern match in left-to-right, top-to-bottom order, but that repeated matches on the same value are allowed (but not guaranteed) to be optimized away into a single comparison. Reiterating that here might be nice.

Reply:
I don't want to focus too much on the specifics of `match`, because this doc is supposed to be generically about how _any_ pattern-matching construct will hook into sum types. Instead I've rephrased to put less emphasis on `match`; does that seem OK?